frontend/lib/resolution/InitResolver.cpp

```diff
@@ -280,6 +280,54 @@ bool InitResolver::isFinalReceiverStateValid(void) {
   return ret;
 }
 
+// Extract domain type information from _instance substitution
+static const DomainType* domainTypeFromSubsHelper(
+    Context* context, const CompositeType::SubstitutionsMap& subs) {
+  auto genericDomain = DomainType::getGenericDomainType(context);
+
+  // Expect one substitution for _instance
+  if (subs.size() != 1) return genericDomain;
+
+  const QualifiedType instanceQt = subs.begin()->second;
+
+  if (auto instance = instanceQt.type()) {
+    if (auto instanceCt = instance->toClassType()) {
+      if (auto instanceBct = instanceCt->basicClassType()) {
+        // Get BaseRectangularDom parent subs for rectangular domain info
+        if (auto baseDom = instanceBct->parentClassType()) {
+          auto& rf = fieldsForTypeDecl(context, baseDom,
+                                       DefaultsPolicy::IGNORE_DEFAULTS);
+          if (baseDom->name() == "BaseRectangularDom") {
+            CHPL_ASSERT(rf.numFields() == 3);
+            QualifiedType rank;
+            QualifiedType idxType;
+            QualifiedType strides;
+            for (int i = 0; i < rf.numFields(); i++) {
+              if (rf.fieldName(i) == "rank") {
+                rank = rf.fieldType(i);
+              } else if (rf.fieldName(i) == "idxType") {
+                idxType = rf.fieldType(i);
+              } else if (rf.fieldName(i) == "strides") {
+                strides = rf.fieldType(i);
+              }
+            }
+
+            return DomainType::getRectangularType(context, instanceQt, rank,
+                                                  idxType, strides);
+          } else if (baseDom->name() == "BaseAssociativeDom") {
+            // TODO: support associative domains
+          } else {
+            CHPL_ASSERT(false && "Not handled!");
```

Comment:
I meant for this to halt because I thought it was all the subclasses of `BaseDom`. I forgot `BaseSparseDom` and will add that, but besides that do you think it's ok to halt?

Reply:
Actually, I suppose in that case it'd make more sense to return the default/error value here (generic domain type).